tsbrowse/preprocess.py

```diff
@@ -0,0 +1,546 @@
+            if num_children[p] == 0:
+                current_num_internal_nodes += 1
+            else:
+                nodes_with_arity[num_children[p]] -= 1
+            num_children[p] += 1
+            nodes_with_arity[num_children[p]] += 1
+            if num_children[p] > current_max_arity:
+                current_max_arity = num_children[p]
+            c = edges_child[e]
+            branch_length = nodes_time[p] - nodes_time[c]
+            current_tbl += branch_length
+        tbl[tree_index] = current_tbl
+        num_internal_nodes[tree_index] = current_num_internal_nodes
+        max_arity[tree_index] = current_max_arity
+        tree_index += 1
+        # print("tree", tree_index, nodes_with_arity)
+
+    return tbl, num_internal_nodes, max_arity
+
+
+def compute_per_tree_stats(ts):
+    """
+    Returns the per-tree statistics
+    """
+    tree_pos = alloc_tree_position(ts)
+    return _compute_per_tree_stats(
+        tree_pos,
+        ts.num_trees,
+        ts.num_nodes,
+        ts.nodes_time,
+        ts.edges_parent,
+        ts.edges_child,
+    )
+
+
+def trees(ts):
+    num_trees = ts.num_trees
+    total_branch_length, num_internal_nodes, max_arity = compute_per_tree_stats(ts)
+
+    # FIXME - need to add this to the computation above
+    mean_internal_arity = np.zeros(num_trees)
+
+    site_tree_index = (
+        np.searchsorted(ts.breakpoints(as_array=True), ts.sites_position, side="right")
+        - 1
+    )
+    unique_values, counts = np.unique(site_tree_index, return_counts=True)
+    sites_per_tree = np.zeros(ts.num_trees, dtype=np.int64)
+    sites_per_tree[unique_values] = counts
+
+    mutation_tree_index = site_tree_index[ts.mutations_site]
+    unique_mutation_values, mutation_counts = np.unique(
+        mutation_tree_index, return_counts=True
+    )
+    mutations_per_tree = np.zeros(ts.num_trees, dtype=np.int64)
+    mutations_per_tree[unique_mutation_values] = mutation_counts
+
+    breakpoints = ts.breakpoints(as_array=True)
+    logger.info("Pre processed trees")
+    return {
+        "left": breakpoints[:-1],
+        "right": breakpoints[1:],
+        "total_branch_length": total_branch_length,
+        "mean_internal_arity": mean_internal_arity,
+        "max_internal_arity": max_arity,
+        "num_sites": sites_per_tree,
+        "num_mutations": mutations_per_tree,
+    }
+
+
+def preprocess(tszip_path, output_path=None):
+    tszip_path = pathlib.Path(tszip_path)
+    if output_path is None:
+        output_path = tszip_path.with_suffix(".tsbrowse")
+    preprocessors = [mutations, nodes, trees, edges, sites]
+    with tqdm(total=2 + len(preprocessors), desc="Processing") as pbar:
+        ts = tszip.load(tszip_path)
+        pbar.update(1)
+
+        tszip.compress(ts, output_path)
+        pbar.update(1)
+
+        # Preprocess the data first so we can error out before writing to the file
+        data = {}
+        with ProcessPoolExecutor() as executor:
```

Comment:
I'm not sure this is worth it - memory usage is going to be an issue and it's going to be hard to control with multiple worker processes. Let's just do it syncronously for now.